# OPENAI's INSTAGRAM FOLLOWERS ANALYSIS

#### Reading file 

In [17]:
with open("finaldata.txt", encoding='utf-8') as f:
    data = f.read()

In [18]:
print(data[:1000])   # First 1000 characters

techie_koramangala
420 posts
9,800 followers
310 following
Arjun M.
Software Engineer
💻 Full-stack @ SaaS startup
☕ Working from cafes in Koramangala
github.com/arjunmakes

blrcoder_girl
150 posts
6,300 followers
120 following
Riya Codes
Tech Creator
👩‍💻 Python | ML | Roadmaps
🎥 Coding Reels + Daily Motivation
linktr.ee/riyacodes

hsr_startup_stories
980 posts
22.4K followers
95 following
Startup Stories HSR
Media
🚀 Interviews with Bangalore founders
🎙️ Weekly startup breakdowns
hsrstartupstories.com

cloud_dev_blr
210 posts
4,500 followers
60 following
Naveen DevOps
Engineer
☁️ AWS | Docker | K8s
📍 Sarjapur
github.com/naveencloud

koramangala_coffee_lover
300 posts
7,900 followers
180 following
Sana B.
Cafe Explorer
☕ Reviewing Bangalore’s hidden coffee spots
📍 Koramangala
youtube.com/@sanacafehunter

ai_bangalore_news
1,340 posts
38K followers
200 following
AI Bangalore
Media
🤖 Daily AI news from India
📢 Startup & research updates
aibanglore.ai/news

techfounder_india
210 posts
19K f

In [19]:
chunks = data.split("\n\n")

#### Creating Structural Dataset

In [20]:
def parse_chunk(chunk):
    chunk = chunk.strip()
    sep_chunk = chunk.split("\n")
    username = sep_chunk[0]
    no_of_posts = int(sep_chunk[1].split(" post")[0].replace(",", ""))
    no_of_followers = float(sep_chunk[2].split(" followers")[0].replace(",", "").replace("K", "").replace("M", ""))
    if("K" in sep_chunk[2]):
        no_of_followers = int(no_of_followers * 1000)
    elif("M" in sep_chunk[2]):
        no_of_followers = int(no_of_followers * 1000000)
    no_of_followers = int(no_of_followers)
    no_of_following = float(sep_chunk[3].split(" following")[0].replace(",", "").replace("K", "").replace("M", ""))
    if("K" in sep_chunk[3]):
        no_of_following = int(no_of_following * 1000)
    elif("M" in sep_chunk[3]):
        no_of_following = int(no_of_following * 1000000)
    no_of_following = int(no_of_following)
    name = sep_chunk[4]
    type_of_page = sep_chunk[5]
    bio = "\n".join(sep_chunk[6:])

    # print(username, no_of_posts, no_of_followers, no_of_following, name, type_of_page, bio)
    return {"username":username, "no_of_posts":no_of_posts, "no_of_followers":no_of_followers, "no_of_following":no_of_following, "name":name, "type_of_page":type_of_page, "bio":bio}


analysis = []

for chunk in chunks:
    analysis.append((parse_chunk(chunk)))

# for chunk in chunks:
#     print(parse_chunk(chunk))




In [21]:
analysis[:3]

[{'username': 'techie_koramangala',
  'no_of_posts': 420,
  'no_of_followers': 9800,
  'no_of_following': 310,
  'name': 'Arjun M.',
  'type_of_page': 'Software Engineer',
  'bio': '💻 Full-stack @ SaaS startup\n☕ Working from cafes in Koramangala\ngithub.com/arjunmakes'},
 {'username': 'blrcoder_girl',
  'no_of_posts': 150,
  'no_of_followers': 6300,
  'no_of_following': 120,
  'name': 'Riya Codes',
  'type_of_page': 'Tech Creator',
  'bio': '👩\u200d💻 Python | ML | Roadmaps\n🎥 Coding Reels + Daily Motivation\nlinktr.ee/riyacodes'},
 {'username': 'hsr_startup_stories',
  'no_of_posts': 980,
  'no_of_followers': 22400,
  'no_of_following': 95,
  'name': 'Startup Stories HSR',
  'type_of_page': 'Media',
  'bio': '🚀 Interviews with Bangalore founders\n🎙️ Weekly startup breakdowns\nhsrstartupstories.com'}]

In [22]:
import json
s = json.dumps(analysis, indent=4)


#### Identify Key Accounts Metrics

In [23]:
maximum_post = analysis[0]['no_of_posts']
max_index = 0
for i in range(0, len(analysis)):
    if maximum_post < analysis[i]['no_of_posts']:
        maximum_post = analysis[i]['no_of_posts']
        max_index = i

print(f"MAX POSTS are {maximum_post} of: ", analysis[max_index], "INDEX_NUMBER: ", max_index, "\n\n")

MAX POSTS are 2300 of:  {'username': 'startuphub_blr', 'no_of_posts': 2300, 'no_of_followers': 45000, 'no_of_following': 120, 'name': 'Startup Hub Bangalore', 'type_of_page': 'Media', 'bio': '🦄 News from the Silicon Valley of India\n📢 Funding alerts, Hiring trends, and Drama\n📩 DM for features\nstartuphub.blr/newsletter'} INDEX_NUMBER:  51 




In [24]:
maximum_followers = analysis[0]['no_of_followers']
max_index_followers = 0
for i in range(0, len(analysis)):
    if maximum_followers < analysis[i]['no_of_followers']:
        maximum_followers = analysis[i]['no_of_followers']
        max_index_followers = i

print(f"MAX FOLLOWERS are {maximum_followers} of: ", analysis[max_index_followers], "INDEX_NUMBER: ", max_index_followers, "\n\n")

MAX FOLLOWERS are 681000 of:  {'username': '_anujsinghal', 'no_of_posts': 1785, 'no_of_followers': 681000, 'no_of_following': 248, 'name': 'Anuj Singhal', 'type_of_page': 'Digital creator', 'bio': 'Managing Editor, CNBC-Awaaz. Most trusted and followed biz anchor in India\nLet’s talk about finance and life\nContact: Singhalanuj1010@gmail.com\nhindi.cnbctv18.com/market-gurukul'} INDEX_NUMBER:  47 




In [25]:
maximum_following = analysis[0]['no_of_following']
max_index_following = 0
for i in range(0, len(analysis)):
    if maximum_following < analysis[i]['no_of_following']:
        maximum_following = analysis[i]['no_of_following']
        max_index_following = i

print(f"MAX FOLLOWINGS are {maximum_following} of: ", analysis[max_index_following], "INDEX_NUMBER: ", max_index_following, "\n\n")

MAX FOLLOWINGS are 890 of:  {'username': 'bangalore_tech_bro', 'no_of_posts': 402, 'no_of_followers': 12500, 'no_of_following': 890, 'name': 'Rahul | HSR Hustler', 'type_of_page': 'Entrepreneur', 'bio': '🚀 Building the next Unicorn in Fintech\n☕ 3rd Wave Coffee addict\n💻 Python | React | AI\n📍 HSR Layout, BLR\nlinktr.ee/rahulbuilds'} INDEX_NUMBER:  49 




In [26]:
Categories = []

for i in range(len(analysis)):
    Categories.append(analysis[i]["type_of_page"])

Categories = set(Categories)
Categories = list(Categories)

print(Categories, "There are", len(Categories), "Categories")

['Full-stack Dev', 'Data Science', 'Recruiter', 'Cafe Nerd', 'Product Manager', 'Education', 'Media', 'Writer', 'Cafe Explorer', 'Digital creator', 'Engineer', 'Lifestyle', 'Developer', 'Blogger', 'Backend Developer', 'Data Scientist', 'Design', 'Artist + Coder', 'Nonprofit organization', 'Software Engineer', 'Comedy', 'Frontend Dev', 'Personal Blog', 'Tech Creator', 'Tech', 'Blog', 'Educator', 'Founder', 'Community', 'Investor', 'Public Figure', 'Food & Drink', 'Coder', 'Entrepreneur'] There are 34 Categories


In [27]:
print("TOP 10 MOST FOLLOWED ACCOUNTS\n")

analysis.sort(key=lambda x: x["no_of_followers"], reverse=True)
for i in range(0, 10):
    print(analysis[i]['username'], "followers = ",  end=" ")
    print(analysis[i]['no_of_followers'])

TOP 10 MOST FOLLOWED ACCOUNTS

_anujsinghal followers =  681000
startuphub_blr followers =  45000
ai_bangalore_news followers =  38000
startup_grind_bangalore followers =  38000
venture_vikram followers =  30000
blrcoders_space followers =  25000
deeptech_india followers =  24000
hsr_startup_stories followers =  22400
hsr_startup_stories followers =  22000
python_blr_jobs followers =  21000


#### Categories distribution analysis

In [28]:
# Count accounts in each category
category_count = {}

for account in analysis:
    category = account["type_of_page"]

    if category in category_count:
        category_count[category] += 1
    else:
        category_count[category] = 1

# Sort by number of accounts (highest to lowest)
sorted_categories = sorted(category_count.items(), key=lambda x: x[1], reverse=True)

# Display
print("Category-wise Account Count")
print("-" * 35)

for category, count in sorted_categories:
    print(f"{category}: {count}")

Category-wise Account Count
-----------------------------------
Community: 33
Media: 12
Blog: 8
Tech: 7
Personal Blog: 5
Education: 4
Lifestyle: 3
Entrepreneur: 2
Food & Drink: 2
Design: 2
Tech Creator: 2
Public Figure: 2
Engineer: 2
Educator: 2
Developer: 2
Digital creator: 1
Investor: 1
Founder: 1
Comedy: 1
Software Engineer: 1
Frontend Dev: 1
Cafe Explorer: 1
Cafe Nerd: 1
Nonprofit organization: 1
Product Manager: 1
Recruiter: 1
Data Science: 1
Full-stack Dev: 1
Artist + Coder: 1
Backend Developer: 1
Writer: 1
Data Scientist: 1
Coder: 1
Blogger: 1


#### Top 10 Account Analysis

In [29]:
top_followers = sorted(analysis, key=lambda x: x["no_of_followers"], reverse=True)

print("TOP 10 FOLLOWED ACCOUNTS")
print("-" * 40)

for i in range(10):
    print(f"{i+1}. {top_followers[i]['username']} : {top_followers[i]['no_of_followers']:,} Followers")

TOP 10 FOLLOWED ACCOUNTS
----------------------------------------
1. _anujsinghal : 681,000 Followers
2. startuphub_blr : 45,000 Followers
3. ai_bangalore_news : 38,000 Followers
4. startup_grind_bangalore : 38,000 Followers
5. venture_vikram : 30,000 Followers
6. blrcoders_space : 25,000 Followers
7. deeptech_india : 24,000 Followers
8. hsr_startup_stories : 22,400 Followers
9. hsr_startup_stories : 22,000 Followers
10. python_blr_jobs : 21,000 Followers


In [30]:
top_following = sorted(analysis, key=lambda x: x["no_of_following"], reverse=True)

print("TOP 10 FOLLOWING ACCOUNTS")
print("-" * 40)

for i in range(10):
    print(f"{i+1}. {top_following[i]['username']} : {top_following[i]['no_of_following']:,} Following")

TOP 10 FOLLOWING ACCOUNTS
----------------------------------------
1. bangalore_tech_bro : 890 Following
2. venture_vikram : 500 Following
3. techfounder_india : 450 Following
4. devlife_bangalore : 450 Following
5. startup_grind_bangalore : 420 Following
6. foodie_in_kora : 400 Following
7. techwalk_blr : 320 Following
8. techie_koramangala : 310 Following
9. blrcoders_space : 300 Following
10. hsr_startup_stories : 300 Following


In [31]:
top_posts = sorted(analysis, key=lambda x: x["no_of_posts"], reverse=True)

print("TOP 10 ACCOUNTS BY POSTS")
print("-" * 40)

for i in range(10):
    print(f"{i+1}. {top_posts[i]['username']} : {top_posts[i]['no_of_posts']:,} Posts")

TOP 10 ACCOUNTS BY POSTS
----------------------------------------
1. startuphub_blr : 2,300 Posts
2. intaglobal : 1,946 Posts
3. _anujsinghal : 1,785 Posts
4. ai_bangalore_news : 1,340 Posts
5. python_blr_jobs : 1,200 Posts
6. startup_grind_bangalore : 1,100 Posts
7. hsr_startup_stories : 980 Posts
8. deeptech_india : 900 Posts
9. foodie_in_kora : 890 Posts
10. koramangala_hustlers : 800 Posts


#### SEARCH ACCOUNT BY USERNAME

In [32]:
search_username = input("Enter Username: ")

found = False

for account in analysis:
    if account["username"] == search_username:
        found = True

        print("\n" + "=" * 45)
        print("ACCOUNT FOUND")
        print("=" * 45)

        print(f"Username   : {account['username']}")
        print(f"Name       : {account['name']}")
        print(f"Followers  : {account['no_of_followers']:,}")
        print(f"Following  : {account['no_of_following']:,}")
        print(f"Posts      : {account['no_of_posts']:,}")
        print(f"Category   : {account['type_of_page']}")
        print(f"Bio        : {account['bio']}")

        break

if not found:
    print("\nUser not found.")


ACCOUNT FOUND
Username   : techie_koramangala
Name       : Arjun M.
Followers  : 9,800
Following  : 310
Posts      : 420
Category   : Software Engineer
Bio        : 💻 Full-stack @ SaaS startup
☕ Working from cafes in Koramangala
github.com/arjunmakes




# Project Summary and Conclusion

## Overview

This project demonstrates the analysis of Instagram profile data using **Pure Python** without relying on external data analysis libraries such as Pandas or NumPy. The dataset was transformed from raw text into structured Python dictionaries, enabling efficient processing through core Python concepts including string manipulation, lists, dictionaries, loops, conditional statements, sorting, and searching.

## Analysis Performed

* Parsed raw Instagram profile data into a structured dataset.
* Identified the accounts with the highest number of posts, followers, and following.
* Counted and analyzed the distribution of account categories.
* Ranked the top 10 accounts based on followers, following, and number of posts.
* Implemented an exact username search system for retrieving complete account information.

## Key Findings

* The dataset contains **34 unique account categories**, indicating a diverse collection of Instagram profiles.
* **Community** is the most common category with **33 accounts**, followed by **Media (12)**, **Blog (8)**, and **Tech (7)**.
* The account **_anujsinghal** has the highest follower count (**681,000**), making it a significant outlier compared to the remaining accounts in the dataset.
* The account **startuphub_blr** has the highest number of posts (**2,300**), suggesting a highly active content publishing strategy.
* The account **bangalore_tech_bro** follows the largest number of profiles (**890**), indicating a broad network compared to other accounts.
* The Top 10 rankings reveal that technology, startup, media, and community-focused accounts dominate the dataset.

## Conclusion

This project demonstrates that meaningful exploratory data analysis can be performed using only Python's built-in data structures and programming constructs. By converting unstructured profile data into structured records, the project successfully extracts useful insights, identifies important metrics, ranks accounts based on multiple criteria, and enables efficient profile searching. The implementation highlights strong problem-solving ability and reinforces core Python programming concepts without relying on third-party data analysis libraries.

